### PR TITLE
add buildrun cancel

### PR DESCRIPTION
**Pull request by gabemontero**
# Changes

This is complimentary to, though not required by, https://github.com/shipwright-io/build/pull/809 in 
addressing  https://github.com/shipwright-io/build/issues/54

/kind feature

# Submitter Checklist

- [ ] Includes tests if functionality changed/was added
- [ ] Includes docs if changes are user-facing
- [ ] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [ ] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

```release-note
Add cancel buildrun option
```

### CODE REVIEW

**Comment by gabemontero** [comment]
FYI I'm going to borrow from the unit tests for openshift's analogous cancel build command from its CLI and incorporate them here.

Could serve as a template for adding tests in for pre-existing code.

**Comment by gabemontero** [comment]
side note:  while `tkn taskrun cancel` provides no args at this time, `oc cancel-build` has some options:

```
Options:
      --dump-logs=false: Specify if the build logs for the cancelled build should be shown.
      --restart=false: Specify if a new build should be created after the current build is cancelled.
      --state=[new,pending,running]: Only cancel builds in this state
```

initial reaction:  all have *some* level of interest but I'm also not 100% of the opinion that we have to take on any of them, at least initially with this PR; if we don't,  wait for feedback as openshift build v1 users migrate to shipwright and/or add a separate issue to note them.

**Comment by gabemontero** [comment]
additional note:  this has been discussed in some internal team meetings, and @otaviof and I have had some discussion on a few implementation alternatives, but for general awareness, I plan on updating this PR with some form of inspection of the buildrun when using the -F / follow options with `shp build run` so that when the pipe on the pod logs is closed, some indication that the buildrun is canceled will be provided.

**Comment by gabemontero** [comment]
I've pushed the change to augment log following if a buildrun is canceled (through in deleted for good measure)

going to 

/hold

since my shipwright-io/build PR has not merged yet, and I wanted to leverage the `IsCanceled` method, so I manually patched that vendor bump in.

once https://github.com/shipwright-io/build/pull/809 merges I'll vendor it in properly.

A snippet of the log:

```shell
$ shp build run buildpack-nodejs-build -F 
Pod 'buildpack-nodejs-build-9k9sp-5bjhb-pod-9nhkx' is in state "Pending"...
Pod 'buildpack-nodejs-build-9k9sp-5bjhb-pod-9nhkx' is in state "Pending"...
Pod 'buildpack-nodejs-build-9k9sp-5bjhb-pod-9nhkx' is in state "Pending"...
[place-tools] 2021/07/12 19:25:17 Copied /ko-app/entrypoint to /tekton/tools/entrypoint
[source-default] 2021/07/12 19:25:18 warning: unsuccessful cred copy: ".docker" from "/tekton/creds" to "/": unable to create destination directory: mkdir /.docker: permission denied
[source-default] {"level":"info","ts":1626117920.4266548,"logger":"git","msg":"ssh","path":"/usr/bin/ssh","version":"OpenSSH_8.0p1, OpenSSL 1.1.1g FIPS  21 Apr 2020"}
[source-default] {"level":"info","ts":1626117920.4312348,"logger":"git","msg":"git","path":"/usr/bin/git","version":"git version 2.27.0"}
[source-default] {"level":"info","ts":1626117920.4454648,"logger":"git","msg":"git-lfs","path":"/usr/bin/git-lfs","version":"git-lfs/2.11.0 (GitHub; linux amd64; go 1.14.4)"}
[source-default] {"level":"debug","ts":1626117920.4455638,"logger":"git","msg":"/usr/bin/git clone --quiet --no-tags --single-branch --depth 1 -- https://github.com/shipwright-io/sample-nodejs /workspace/source"}
[source-default] {"level":"debug","ts":1626117921.464589,"logger":"git","msg":"/usr/bin/git -C /workspace/source submodule update --init --recursive --depth 1"}
[source-default] {"level":"debug","ts":1626117921.5359373,"logger":"git","msg":"/usr/bin/git -C /workspace/source rev-parse --verify HEAD"}
[prepare] GGMsleep
[build-and-push] rpc error: code = NotFound desc = an error occurred when try to find container "2485a45b709fad5d4d8433214da6a14395c9608af000222f9ed6050a1dc13d71": not found
[prepare] rpc error: code = NotFound desc = an error occurred when try to find container "9f2c014a1041d191d2e112cd20733a83ffa2d101b5a05a892f05caab4346069c": not found
BuildRun 'buildpack-nodejs-build-9k9sp' has been canceled.
$
```

**Comment by gabemontero** [comment]
ok

/hold cancel

as I've vendored in the recently merged shipwright-io/build changes for cancelling builds 

any takers on code review @coreydaley @otaviof @alicerum @HeavyWombat ?

**Comment by coreydaley** [comment]
I'm on the case!

**Comment by coreydaley** [comment]
I'm on the case!

File: pkg/shp/cmd/build/run.go [hunk 1]
```diff
@@ -6,6 +6,7 @@ import (
 	"time"
 
 	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	buildclientset "github.com/shipwright-io/build/pkg/client/clientset/versioned"
```
Comment (coreydaley) [inline pkg/shp/cmd/build/run.go:9]: The below `k8s.io` etc imports should be split out into their own blocks.

File: pkg/shp/cmd/build/run.go [hunk 2]
```diff
@@ -137,11 +154,17 @@ func (r *RunCommand) Run(params *params.Params, ioStreams *genericclioptions.IOS
...
+	if !r.follow {
 		fmt.Fprintf(ioStreams.Out, "BuildRun created %q for build %q\n", br.GetName(), r.buildName)
 		return nil
 	}
 
+	r.buildRunName = br.Name
+	r.shpClientset, err = params.ShipwrightClientSet()
+	if err != nil {
+		return err
+	}
```
Comment (coreydaley) [inline pkg/shp/cmd/build/run.go:166]:
```suggestion
	if r.shpClientset, err = params.ShipwrightClientSet();  err != nil {
		return err
	}
```

File: pkg/shp/cmd/buildrun/cancel.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// Complete fills in data provided by user
+func (c *CancelCommand) Complete(params *params.Params, args []string) error {
+	c.name = args[0]
+
+	return nil
+}
+
+// Validate validates data input by user
+func (c *CancelCommand) Validate() error {
+	return nil
```
Comment (coreydaley) [inline pkg/shp/cmd/buildrun/cancel.go:48]: Can a BuildRun name be any length and contain any characters? Or does it need to adhere to some kind of label rules or regexp?

File: pkg/shp/cmd/buildrun/cancel.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+// Run executes cancel sub-command logic
+func (c *CancelCommand) Run(params *params.Params, ioStreams *genericclioptions.IOStreams) error {
+	brr := resource.GetBuildRunResource(params)
+
+	br := &buildv1alpha1.BuildRun{}
+	err := brr.Get(c.cmd.Context(), c.name, br)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve BuildRun %s: %s", c.name, err.Error())
+	}
```
Comment (coreydaley) [inline pkg/shp/cmd/buildrun/cancel.go:59]:
```suggestion
	if err := brr.Get(c.cmd.Context(), c.name, br); err != nil {
		return fmt.Errorf("failed to retrieve BuildRun %s: %s", c.name, err.Error())
	}
```

File: pkg/shp/cmd/buildrun/cancel.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	//TODO replace with br.IsDone() when that is available and vendored in
+	cond := br.Status.GetCondition(buildv1alpha1.Succeeded)
+	if cond != nil && cond.GetStatus() != corev1.ConditionUnknown {
+		return fmt.Errorf("failed to cancel BuildRun %s: BuildRun has already finished execution", c.name)
```
Comment (coreydaley) [inline pkg/shp/cmd/buildrun/cancel.go:63]:
```suggestion
		return fmt.Errorf("failed to cancel BuildRun %s: execution has already finished", c.name)
```

File: pkg/shp/cmd/buildrun/cancel.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	//TODO use constant when vendor in api changes
+	if err := brr.Patch(c.cmd.Context(), c.name, "replace", "/spec/state", "BuildRunCanceled"); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(ioStreams.Out, "BuildRun cancelled '%v'\n", c.name)
```
Comment (coreydaley) [inline pkg/shp/cmd/buildrun/cancel.go:71]:
```suggestion
	fmt.Fprintf(ioStreams.Out, "BuildRun '%s' successfully cancelled\n", c.name)
```

File: pkg/shp/util/crud.go [hunk 7]
```diff
@@ -83,3 +85,20 @@ func ListObjectWithOptions(ctx context.Context, resource dynamic.ResourceInterfa
...
+	Value string `json:"value"`
+}
+
+func PatchObject(ctx context.Context, resource dynamic.ResourceInterface, name, op, path, value string) error {
+	payload := []patchStringValue{{
+		Op:    op,
+		Path:  path,
+		Value: value,
+	}}
+	data, _ := json.Marshal(payload)
```
Comment (coreydaley) [inline pkg/shp/util/crud.go:101]: I think it would be worth checking the error here

**Review by coreydaley — CHANGES_REQUESTED** [review]
Are there any tests that should go along with this for the new command?

File: pkg/shp/cmd/buildrun/cancel.go (hunk 6 above)
Comment (gabemontero) [reply pkg/shp/cmd/buildrun/cancel.go:71]: Sascha wanted to go with the US `canceled` vs. UK `cancelled`.

I'll fix that and included the rest of your suggestion.

Comment (coreydaley) [reply pkg/shp/cmd/buildrun/cancel.go:71]: My comment was just to change the wording, use whichever spelling.

File: pkg/shp/cmd/buildrun/cancel.go (hunk 3 above)
Comment (gabemontero) [reply pkg/shp/cmd/buildrun/cancel.go:48]: that's the type of validation to be done on the generated code side in shipwright-io/build ... leaving this as is

as a point of comparison, while there are analogous Run and Complete semantics over in openshift/oc, there is not "Validate" one.

Looking at the other verbs here, I see one instance where we check for the name being empty, but it seems like the arg check should over that

Feels like a separate cli issue of (what should we do with Validate / should we keep Validate)

Not pulling that into this PR

**Comment by gabemontero** [comment]
> Are there any tests that should go along with this for the new command?

Yeah I'm hesitant to devise tests for pkg/shp/cmd/* until we completely move off of the dynamic client and onto static clients per https://github.com/shipwright-io/cli/issues/24

That said we may need to get on that before we add to many more verbs

@otaviof @alicerum FYI

**Comment by gabemontero** [comment]
Otherwise I've pushed updates for comments from @coreydaley (who is on PTO) for the remaining items from him I'm in line with.

/assign @alicerum 
/assign @otaviof 

Can between the two of you all take a look at this one and see if you all have anymore input.

Once I get agreement from you all on the raw content, I can squash / sort commits as needed and we can officially l-g-t-m

thanks

**Review by otaviof — APPROVED** [review]
Just had my first pass at it, it looks good, thanks Gabe!

/lgtm

**Comment by gabemontero** [comment]
awesome thanks @otaviof !

**Pull request merged**